### ISSUE TITLE
ABI break in 3.0.3

### ISSUE THREAD

**Issue by bbhtt** [libjpeg-turbo/libjpeg-turbo#773]
**Have you searched the existing issues (both open and closed) in the libjpeg-turbo issue tracker to ensure that this bug report is not a duplicate?**

Yes

**Does this bug report describe one of the [two known and unsolvable issues with the JPEG format](https://libjpeg-turbo.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf)?**

No

**Clear and concise description of the bug:**

Updating from 3.0.2 to 3.0.3 shows a backwards ABI/API break. Is this expected/public API?

```
        ┌──────────────────────────┐
        │ ABI Break: libjpeg.so.62 │
        └──────────────────────────┘
        Functions changes summary: 0 Removed, 0 Changed, 0 Added function
        Variables changes summary: 1 Removed, 0 Changed, 0 Added variable
        1 Removed variable:
          [D] 'const char* const jpeg_std_message_table[131]'    {jpeg_std_message_table@@LIBJPEG_6.2}
```

**Steps to reproduce the bug (using *only* libjpeg-turbo):** N/A

**Image(s) needed in order to reproduce the bug (if applicable):**N/A

**Expected behavior:**

There should not be a backwards ABI break in a minor release

**Observed behavior:** ABI break.

**Platform(s) (compiler version, operating system version, CPU) on which the bug was observed:** Linux

**libjpeg-turbo release(s), commit(s), or branch(es) in which the bug was observed (always test the tip of the main branch or the latest [stable pre-release](https://libjpeg-turbo.org/DeveloperInfo/PreReleases) to verify that the bug hasn't already been fixed):** 3.0.3

**If the bug is a regression, the specific commit that introduced the regression (use `git bisect` to determine this):** N/A

**Additional information:** N/A

**Comment by dcommander** [comment]
Referring to #767 and #763, that symbol was never exposed in the libjpeg API.  It was historically exposed in the libjpeg link library but only on Un*x platforms or when using the static link libraries on Windows.  (The DLL definition file prevented the symbol from being exposed in the Windows DLL.)  The only documentation of it was in a code comment from the early 1990s left by Tom Lane.  libjpeg was originally intended to be customized at compile time to suit specific application needs, which is why things such as the layout of components in the `JCS_RGB` colorspace, the JPEG data precision, and whether or not to include certain features were configurable with compile-time macros.  The exposure of `jpeg_std_message_table[]` in the link libraries was a relic of that compile-time customization.  (Bear in mind that concepts like a fixed API and shared library versioning weren't really a thing when libjpeg was first created.)  When Tom Lane stepped back from the IJG in the late 1990s, the libjpeg v6b API/ABI remained fixed for about 10 years, and it became sort of a de facto standard that later projects (including mine) had to emulate because of the critical mass of applications that relied upon it.  But only the public API defined in the public headers (**jpeglib.h**, etc.) is part of that backward compatibility contract.  It has always been the case that, if a calling application uses internal functions or symbols exposed in the library (including those defined in **jpegint.h**), then it is responsible for version management (which means either remaining with a fixed version of libjpeg-turbo or managing any changes to the internal functions/symbols that the application uses.)

RawTherapee is the only application known to be affected by this.  Per above, it was really their responsibility to do version management regarding their use of internal symbols, but even so, I wouldn't have made the symbol private had I known that it would break their build (or, at least, I would have deferred that until a major release so it could go through a beta cycle.)  However, what's done is done.  The symbol could not be relied upon by cross-platform applications anyhow, so libjpeg-turbo 3.0.3 is actually more cross-platform-consistent than previous versions were.  If I reintroduced the symbol in libjpeg-turbo 3.0.4, it would create potential confusion, and since RawTherapee has already worked around the issue, I think it's best to keep things the way they are.  I would argue that, if a symbol is not part of the public API, it isn't part of the public ABI either.  Applications wouldn't know how to use the `jpeg_std_message_table[]` symbol unless they read the code, and if the proper functioning of an application relies upon a non-public symbol and a buried code comment in a library, then that isn't exactly robust.

I am open to being convinced otherwise, but my understanding is that the ABI is the binary interface for the API.  If something isn't defined as part of the API, then it isn't part of the ABI either.  If I did re-introduce the symbol, I would have to make sure it worked on Windows as well (which it never has in the 14 years that libjpeg-turbo has been a project.)  Personally, I think it's time to jettison that legacy, because there is already a more robust, cross-platform, and documented way of accessing the same table.
